Serialization 2.0

lixin-wei opened the pull request:
Add REP for Serialization 2.0

simon-mo commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+### General Motivation
+Current Ray's serialization has some issues:
+
+1. Doesn't support out-of-band serialization in Java worker. Cause some perfomance issues in Java worker.
```
define out-of-band

raulchen replied:
@lixin-wei you can add some concrete examples. For example, there was a requirement for zero-copy reading Arrow data in Java.

raulchen replied:
Also, you can talk about how Python currently supports OOB with Pickle. You can link this [PEP](https://peps.python.org/pep-0574/).

simon-mo commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+Current Ray's serialization has some issues:
+
+1. Doesn't support out-of-band serialization in Java worker. Cause some perfomance issues in Java worker.
+1. Hard to add a new serializer for a specific class.
```
class or format? class here is something like user defined or third party class; format is the serialized bytes format

lixin-wei replied:
I think class makes more sense here. From the users' view, they want to add a serializer for a specific `class`. Or maybe `type`?

simon-mo commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+## Design and Architecture
+### User-level API
+#### Register Serializer
+Firstly, if a user wants to implement custom serialization, he should register his serializer to Ray.
```
```suggestion
Firstly, if users wants to implement custom serialization, they should register their serializer to Ray.
```

simon-mo commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,162 @@
+Ray.registerSerializer("ArrowTable", ArrowTable.class, new ArrowTableSerializer());
+Ray.registerSerializer("Protobuf", Protobuf.class, new ProtobufSerializer());
+```
+Ray will maintain a map from ClassID to Serializer in memory. This relationship won't be persisted, which means you need to register them again when the process is restarted.
````
how does this work for worker process?

raulchen replied:
I think, as the first step, this API is only used by Ray core. That means we can simply hard code them without needing persistence. 
If we are to expose this API to users, persistence is needed.

lixin-wei replied:
Yeah just as Hao said, we'll hard code all default serializers into the worker's code.

simon-mo commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,162 @@
+```
+Ray will maintain a map from ClassID to Serializer in memory. This relationship won't be persisted, which means you need to register them again when the process is restarted.
+#### Implement Serializer
+Then, he needs to implement this custom serializer, implements the interface shown below:
````
```suggestion
Then, the user needs to implement this custom serializer, implements the interface shown below:
```

simon-mo reviewed (commented):
some edits suggestion

ericl commented:
Can you add some discussion on how pickle5 support fits into this framework? For example:
- Can pickle5 be registered as the lowest priority serializer for generic Python "object" types?
- How nested serialization works (e.g., Arrow table contained within a in pickle5 object).

It would be awesome to get rid of pickle5 as a special case!

suquark commented:
Yes, I am also thinking of setting pickle5 as a special case! Basically, Ray could have different serialization protocols, some are specialized for a language, e.g., python; but some are more general e.g., pyarrow.

Besides all these serialization frameworks, Ray should have a **general** protocol for basic and buffer-based objects
* it supports basic types like dicts, lists, bytes, ints, string for all languages, 
* it supports buffer-based objects like numpy ndarray, pytorch tensor, pyarrow table; they can be encoded as a dict + buffers.
* it supports GPU and other buffers
* this serialization protocol allows us to transfer objects with zero-copy across languages and across frameworks (numpy -> pytorch -> jax tensor -> ...), which cannot be supported by pyarrow

From the lesson of Arrow project, it is important to have your own standard for data format / serialization, and if we support buffer-based objects with our protocol, it can integrates well with our object store, so we can do something more interesting (e.g., we can move a numpy array to GPU as a pytorch tensor in our object store; because we know our data format in object store, we can do it without the help of numpy and pytorch; or we can spilling a buffer to S3 without touch serialization)

suquark commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+    def serialize(self, object: MyClass) -> RaySerializationResult:
+        pass
+    def deserialize(self, in_band_data: bytes,
+                    out_of_band_data: List[bytes]) -> MyClass:
```
it should be something like buffers for zero copying?

lixin-wei replied:
Yes, thank you for the catching! I translated this code from Java. My mistake. I'll fix this.

suquark commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,162 @@
+### Final Buffer Protocol
+```
++---------------+----------------+-------------------------------------
+|  ClassIDHash  |  in-band data  |   … oob-buf0 … oob-buf1 … oob-buf2 …
````
how about also treating in-band data also as a buffer?

lixin-wei replied:
Fixed, thanks.

lixin-wei replied:
After thinking again, I think in-band buffer shouldn't be a no-copy buffer. Since it will cause data-sharing. If users want to do zero-copy just use out-of-band buffers.

suquark replied:
Both in-band/out-band buffer shouldn't be a no-copy buffer due to data copying. But they can be a readonly buffers. My concern with `bytes` it that in Python, converting to `bytes` usually results in extra copy, but I did know some ways to turn readonly buffers to bytes, so I am ok with `bytes` for in-band data.

suquark commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,162 @@
+```python
+class RaySerializationResult:
+    def __init__(self):
+        self.in_band_buffer: bytes = None
````
maybe also a list of Ray objectrefs? Same in the deserialization API.

lixin-wei replied:
Why Ray object refs? I can't get your point. Could you please say more about this?

suquark replied:
I think this is maybe out of context and orthogonal for this REP. Let me resolve this.

raulchen commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+
+1. Doesn't support out-of-band serialization in Java worker. Cause some perfomance issues in Java worker.
+1. Hard to add a new serializer for a specific class.
+1. Doens't support cross-language serialization for custom classes.
```
Besides custom classes. There are other issues as well. For example: 1) type loss (Java's int/short/byte); 2) support for maps. Also, let's put more examples here.

raulchen commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+1. Provide pluggable ways for users to implement custom serialization, including:
+   1. Cross-language serialization.
+   1. Out of band serialization and other optimizations.
+2. Unify the current serialization code path to this new pluggable design. Make code cleaner.
```
Also, provide a unified interface across different languages.

raulchen commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,162 @@
+Let's take Python as an example. Other languages will be similar.
+```python
+class MySerializer(RaySerializer):
+    def serialize(self, object: MyClass) -> RaySerializationResult:
````
If we return `RaySerializationResult` (which includes the in-bind buffers), we'll need 2 copies: one from user object to RaySerializationResult and another one to append RaySerializationResult's buffers to the final result.
To avoid this copy, we can pass in a "InBindBufferWriter" that can directly write user object data to the final result.

lixin-wei replied:
In this case why not just use `RayOutOfBandBuffer`?

raulchen commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+        pass
+    # dest is a byte array with size == get_size()
+    def write_to(self, dest: bytes):
+        pass
```
We can add a full example for a concrete type to let people better understand how each class works together.

lixin-wei replied:
Added

raulchen commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+Unify all existing object serialization to this API.
+
+1. Implement internal object serializer, such as RayError, ActorHandle, etc.
+1. By default, if no serializer is registered, use current serializer as the fallback. For example, pickle5 in Python, FST in Java. **In this case, cross-language serialization is disabled.**
```
Can we avoid this regression? For example, you can register Msgpack as the serializer for the primitive types.

raulchen commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,162 @@
+## Test Plan and Acceptance Criteria
+Unit tests for core components.
+Compatibility is covered by CI.
+Additional unit tests for individual serializers in the future.
```
Also need performance benchmarks as well.

lixin-wei commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,162 @@
+        self.in_band_buffer: bytes = None
+        self.out_of_band_buffers: List[RayOutOfBandBUffer] = None
+```
+`OutOfBandBuffer` is for advanced users. They can use this to do out-of-band optimization. Ray will call `OutOfBandBuffer#write` only once when constructing the final buffer sent to the network.
````
Example added.

lixin-wei commented:
@raulchen @suquark @simon-mo @jjyao All comments resolved/replied. Added many examples to help you to understand. Would you please have a look again?

lixin-wei commented:
@ericl @suquark For the nested serialization, I added more discussion in the REP. Pasted below.
My main idea is not to support nested serialization in fallback serializer now. What do you think?

--------------------------

### Nested Serialization

Let's take `List[bytearray]` as an example to see how to handle nested types. We've already implemented a serializer for `bytearray` above.

Note that in real world we might simply wrap Msgpack to do serialization for common classes. Here we rebuild the wheel only to show how it works.

```python
class ListSerializer(RaySerializer):

    def serialize(self, list_object: List[str]) -> RaySerializationResult:
        result = RaySerializationResult()
        result.in_band_buffer = bytearray()
        # Total element count.
        result.in_band_buffer.extend(to_bytes(len(list_object))))
        for element in list_object:
            element_res = ray.serialize(element)
            # in band buffer length
            result.in_band_buffer.extend(to_bytes(len(element_res.in_band_buffer)))
            # in band buffer
            result.in_band_buffer.extend(element_res.in_band_buffer)
            # append out of band buffer
            result.out_of_band_buffers += element_res.out_of_band_buffers
        return result

    def deserialize(self, serialization_result: RaySerializationResult, oob_offset: int = 0) -> Tuple[List, int]:
        in_band = serialization_result.in_band_buffer
        oob = serialization_result.out_of_band_buffers

        result = []
        pos = 0
        # read element count
        count, pos = get_int(in_band, pos)
        for i in range(count):
            # read element size
            element_size, pos = get_int(in_band, pos)
            # read element in-band data
            element_in_band_data = in_band_buffer[pos: pos + element_size]
            pos = pos + element_size
            # nested serialization here
            element, oob_offset = ray.deserialize(element_in_band_data, oob, oob_offset)
            result.append(element)
        return (result, oob_offset)
```

### Fallback Serializer

By default, if no serializer is registered, use serializer in current Ray's version as the fallback. For example, pickle5 in Python, FST in Java.
**In this case, cross-language serialization is disabled,** since the protocol are different between different frameworks.

**Also, nested serialization is disabled in this case.** Because the fallback third-party serialization framework is out of our control.
It's possible to modify the third-party framework to make it use Ray's serializer, but I think it's a massive work and not worth it. Maybe a follow-up work item.

lixin-wei commented:
@raulchen @suquark @simon-mo @jjyao All comments resolved/replied. Added many examples to help you to understand. Would you please have a look again?

lixin-wei commented:
@ericl @suquark For the nested serialization, I added more discussion in the REP. Pasted below.
My main idea is not to support nested serialization in fallback serializer now. What do you think?

--------------------------

### Nested Serialization

Let's take `List[bytearray]` as an example to see how to handle nested types. We've already implemented a serializer for `bytearray` above.

Note that in real world we might simply wrap Msgpack to do serialization for common classes. Here we rebuild the wheel only to show how it works.

```python
class ListSerializer(RaySerializer):

    def serialize(self, list_object: List[bytearray]) -> RaySerializationResult:
        result = RaySerializationResult()
        result.in_band_buffer = bytearray()
        # Total element count.
        result.in_band_buffer.extend(to_bytes(len(list_object))))
        for element in list_object:
            element_res = ray.serialize(element)
            # in band buffer length
            result.in_band_buffer.extend(to_bytes(len(element_res.in_band_buffer)))
            # in band buffer
            result.in_band_buffer.extend(element_res.in_band_buffer)
            # append out of band buffer
            result.out_of_band_buffers += element_res.out_of_band_buffers
        return result

    def deserialize(self, serialization_result: RaySerializationResult, oob_offset: int = 0) -> Tuple[List, int]:
        in_band = serialization_result.in_band_buffer
        oob = serialization_result.out_of_band_buffers

        result = []
        pos = 0
        # read element count
        count, pos = get_int(in_band, pos)
        for i in range(count):
            # read element size
            element_size, pos = get_int(in_band, pos)
            # read element in-band data
            element_in_band_data = in_band_buffer[pos: pos + element_size]
            pos = pos + element_size
            # nested serialization here
            element, oob_offset = ray.deserialize(element_in_band_data, oob, oob_offset)
            result.append(element)
        return (result, oob_offset)
```

### Fallback Serializer

By default, if no serializer is registered, use serializer in current Ray's version as the fallback. For example, pickle5 in Python, FST in Java.
**In this case, cross-language serialization is disabled,** since the protocol are different between different frameworks.

**Also, nested serialization is disabled in this case.** Because the fallback third-party serialization framework is out of our control.
It's possible to modify the third-party framework to make it use Ray's serializer, but I think it's a massive work and not worth it. Maybe a follow-up work item.

ericl commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,263 @@
+            element_in_band_data = in_band_buffer[pos: pos + element_size]
+            pos = pos + element_size
+            # nested serialization here
+            element, oob_offset = ray.deserialize(element_in_band_data, oob, oob_offset)
```
I'm a bit confused what oob_offset is used for in this example. It never seems to be used for deserialization at all, just passed through and stays as 0.

Can you modify the example so it actually demonstrates why you need oob_offset?

ericl commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,263 @@
+By default, if no serializer is registered, use serializer in current Ray's version as the fallback. For example, pickle5 in Python, FST in Java.
+**In this case, cross-language serialization is disabled,** since the protocol are different between different frameworks.
+
+**Also, nested serialization is disabled in this case.** Because the fallback third-party serialization framework is out of our control.
```
This seems like a serious flaw. Our pickle5 already supports out of band buffers today. Why can't we implement Pickle5 as a Ray custom serializer? I think the code modification just involves changing the serialization hooks.

ericl reviewed (commented):
This is starting to look very solid. I have only two blocking comments:
1. The example of oob_offset is confusing. It needs to be clear to the reader what this offset is used for, and e.g., why we can't use an array view wrapper instead.
2. We absolutely have to support pickle5 as a custom serializer--- otherwise this proposal will be adding to the complexity of our serialization stack instead of simplifying it. It's not a "massive effort", the code is just the stuff in this file: https://github.com/ray-project/ray/blob/7b0c77dd38ab4943f64e29c51c3427085e27b239/python/ray/serialization.py#L176

jovany-wang commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,263 @@
+**In this case, cross-language serialization is disabled,** since the protocol are different between different frameworks.
+
+**Also, nested serialization is disabled in this case.** Because the fallback third-party serialization framework is out of our control.
+It's possible to modify the third-party framework to make it use Ray's serializer, but I think it's a massive work and not worth it. Maybe a follow-up work item.
```
It seems most serialization frameworks(at least FST, pickle5) support the ability to register custom serializers for specific types. We don't need to change the 3rd party frameworks, just to register.

lixin-wei replied:
Yeah for FST it does have such API, but I can't find one for pickle5, do you know how to do this in pickle5?

lixin-wei commented:
> We absolutely have to support pickle5 as a custom serializer--- otherwise this proposal will be adding to the complexity of our serialization stack instead of simplifying it. It's not a "massive effort", the code is just the stuff in this file: https://github.com/ray-project/ray/blob/7b0c77dd38ab4943f64e29c51c3427085e27b239/python/ray/serialization.py#L176

@ericl The code your mentioned seems only can make pickle5 as Ray's custom serializer. But I don't have a good idea to make Pickle5 use Ray's serializer, do you have any idea about this?

I agree that it's easy to make pickle5 as Ray's custom serializer. The OOB buffers of pickle5 can be easily integrated with Ray. Something likes the code below:

```python
class Pickle5FallbackSerializer(RaySerializer):
    def serialize(self, instance) -> RaySerializationResult:
        oob_buffers = []
        in_band = pickle5.dumps(instance, buffer_callback=oob_buffers.append)
        return RaySerializationResult(in_band, oob_buffers)


    def deserialize(self, serialization_result: RaySerializationResult,
                    oob_offset: int = 0):
        in_band = serialization_result.in_band_buffer
        oob = serialization_result.out_of_band_buffers
        return pickle5.loads(in_band, buffers=oob)
```

But it seems there isn't a easy way to make pickle5 use Ray's serializer.

lixin-wei commented:
OK I've found one possible solution: https://docs.python.org/3/library/pickle.html#dispatch-tables

Let me think more about the detailed implementation.

ericl commented:
> OK I've found one possible solution: https://docs.python.org/3/library/pickle.html#dispatch-tables

Yup, in serialization.py:SerializationContext we are already registering a bunch of custom serialization handlers for ObjectRef, ActorHandle, etc. It would not be hard to register other types of serializers in the same way.

raulchen commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,263 @@
+
+#### Register Serializer
+
+Firstly, if users wants to implement custom serialization, they should register their serializer to Ray.
```
```suggestion
Firstly, if users want to implement custom serialization, they should register their serializers to Ray.
```

raulchen commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,263 @@
+
+```python
+# In Python
+ray.register_serializer("ArrowTable", type(arrow_table_obj), ArrowTableSerializer())
````
```suggestion
ray.register_serializer("ArrowTable", ArrowTable, ArrowTableSerializer())
```

lixin-wei commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,263 @@
+
+* Solve issue 4 immediately.
+* Solve issue 3 by providing build-in serializers for commonly used classes.
+* Solve issue 2 by register different serializer for `int` and `short`.
```
Yes, implementing different serializers for `int` and `short` means saving different type information in the buffer.

lixin-wei commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,263 @@
+Current Ray's serialization has some issues:
+
+1. Doesn't support [out-of-band(OOB) data](https://en.wikipedia.org/wiki/Out-of-band_data). So we can't do zero-copy reading/writing. There was a requirement for zero-copy reading Arrow data in Java, but we couldn't achieve it because of this.
+2. Type loss in cross-lang serialization. e.g. `short` will become `int` from Java to Python.
```
fixed

lixin-wei commented:
@ericl @raulchen  All comments fixed. 
1. I replaced the confusing `oob_offset` with an iterator.
2. More discussion was added for the nested serialization in the fallback serializer.

Would you please have a look again?

lixin-wei commented:
@suquark Would you please double-check this proposal? I'm new to this part and I'm afraid I lost anything :)

ericl commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,300 @@
+
+```python
+# Pseudo code
+def register_serializer(class_name: str, class_type: type, serializer):
````
Instead of special casing pickle5, could we add this to the serialization API? Basically, notify each serializer if another serializer is registered after it. That way the pickle5 custom serializer can be notified that ArrowTable, etc. have custom serializers.

ericl replied:
Or do you want to make it so that only the fallback serializer can see all serializers?

lixin-wei replied:
> Instead of special casing pickle5, could we add this to the serialization plugin API? 

@ericl Yes I agree with you.

> Or do you want to make it so that only the fallback serializer can see all serializers?

Yes I think only the fallback serializer. Notify all serializers seems like an overkill.
Committed, would you please have a look?

ericl replied:
It seems good to me. In the ideal case, we would support nesting any kind of serializers, but I can't think of a concrete use case right now, and this also seems fairly easy to add later should we need it.

lixin-wei commented:
Do you have any other comments? Can we merge this if not? :)

ericl commented:
I think we can do a committer vote soon. Let me discuss offline with Simon and Ryans to see if there are any other issues we might want to address first.

rkooo567 commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,314 @@
+    def serialize(self, object: MyClass) -> RaySerializationResult:
+        pass
+    def deserialize(self, in_band_buffer: bytes,
+                    oob_iterator: Iterator[memoryview]) -> MyClass:
```
Consider using the same name as the field in `RaySerializationResult ` (i.e., out_of_band_buffers)?

lixin-wei replied:
It's an iterator, not buffers. How about `out_of_band_buffers_iterator`?

rkooo567 commented on reps/2022-04-07-serialization.md:
````diff
@@ -0,0 +1,314 @@
+```python
+class RaySerializationResult:
+    in_band_buffer: bytes
+    out_of_band_buffers: Iterable[memoryview]
````
Should we mark it Optional since it is not needed for all serialization?

lixin-wei replied:
Sounds good!

rkooo567 commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,314 @@
+def register_fallback_serializer(serializer, register_callback=None):
+    pass
+
+def register_serializer(class_name: str, class_type: type, serializer):
```
Is it necessary to split class name & type? Can't one infer from the other?

lixin-wei replied:
The class_name is a cross-lang ID, it's specified by users. We can't generate it from class_type.

rkooo567 commented on reps/2022-04-07-serialization.md:
```diff
@@ -0,0 +1,314 @@
+
+By default, if no serializer is registered, use serializer in current Ray's version as the fallback. For example, pickle5 in Python, FST in Java.
+
+We'll register a fallback serializer to Ray. For example, pickle5 fallback serializer will be something like the following, all OOB buffers of pickle5 will be wrapped to Ray's OOB buffer.
```
What does it exactly mean by fallback here? Does it mean it's used by unregistered classes or is it used when failure happens from the original serializer?

lixin-wei replied:
it's used by unregistered classes. I changed it to "language default serializer" for clarity.

ericl reviewed (commented):
Should oob_iterator be `oob_buffers: Map[id: int, memoryview]`? This is more explicit, since an iterator can be confusing.

Deserializers can lookup up their buffer by their unique id. During serialization, we can provide a way to generate new ids (`serialize(object, cur_out_of_buffer_id) -> (bytes, Map[id, memoryview])`)

ericl commented:
Summarizing some minor questions from a discussion with committers (cc @clarkzinzow @edoakes @stephanie-wang @suquark @simon-mo @pcmoritz ):
1. Can we also register our ObjectRef/ActorHandle custom serialization as custom serializer hooks? 
2. Will nested serialization of ObjectRef work if the ObjectRef is within a pickle5 buffer within a msgpack buffer? It seems there might be a problem if only the "fallback" serializer can know about registered serializers, since in this case msgpack is the fallback serializer, not pickle5. Or are both "fallback" serializers?
3. Can we flesh out how the registration process works for serializers? How do worker nodes get the registration info?
4. Can we rename "fallback serializer" to "language default serializer" for clarity?
5. In the initial version, could we mark the serialization plugin API as private or DeveloperAPI?

Beyond these questions, the main concern was proving the proposed plugin cleanup is actually simplifying our serialization architecture and not regressing any existing use cases.

Is it possible to create a prototype refactoring the current Python serialization code and demonstrating that it passes all unit and nightly tests? Showing the proposal is simplifying the code would erase any doubts about the benefits of the approach.

lixin-wei commented:
> Should oob_iterator be oob_buffers: Map[id: int, memoryview]? This is more explicit, since an iterator can be confusing. This is the way the pickle5 serializer works right now I believe.

Sounds good, it's more explicit.

> Can we also register our ObjectRef/ActorHandle custom serialization as custom serializer hooks?

Sure we can.

> Can we flesh out how the registration process works for serializers? How do worker nodes get the registration info?

I mentioned this in the "Register Serializer" paragraph: As the first step, this API is only used by Ray core. That means we can simply hard code them without needing persistence. If we are to expose this API to users, we can consider to add the persistance feature in the future.


> Can we rename "fallback serializer" to "language default serializer" for clarity?

Done.

> In the initial version, could we mark the serialization plugin API as private or DeveloperAPI?

Sure we can.

lixin-wei commented:
> Will nested serialization of ObjectRef work if the ObjectRef is within a pickle5 buffer within a msgpack buffer? It seems there might be a problem if only the "fallback" serializer can know about registered serializers, since in this case msgpack is the fallback serializer, not pickle5. Or are both "fallback" serializer?

@ericl I can't understand this case, could you please say more about when will this case happen? Why do we need msgpack here? Isn't pickle5 enough?

IICU, maybe you were talking about nested serialization in cross-lang case? As I mentioned in the REP, unfortunately, it's not supported.

```Markdown
**Cross-language serialization is disabled in this case.** since the protocol are different between different frameworks.
If users want to do cross-language serialization, unfortunately, they still need to implement a simple wrapper serializer.

This can be mitigated if we can find a full-featured cross-language serialization library as the default serializer. But we can't find one by now. Maybe we can use [Fury](https://docs.google.com/document/d/1nrKrXnyRqiIQqLV1P6i3t6TXQEwduoyLeHLMT2DV-fc/edit?usp=sharing) in the future after it's open-source.
```

lixin-wei commented:
> Beyond these questions, the main concern was proving the proposed plugin cleanup is actually simplifying our serialization architecture and not regressing any existing use cases.

@ericl To clarify, this refactoring might not make code *simpler*, it might add more abstraction. The main objective is to provide a standard way to implement custom cross-lang serializers.

> Is it possible to create a prototype refactoring the current Python serialization code and demonstrating that it passes all unit and nightly tests? Showing the proposal is simplifying the code would erase any doubts about the benefits of the approach.

Sure I can submit a prototype PR, I'm confident that it will pass all tests. But the code might not be simpler as I mentioned above.

ericl commented:
> @ericl I can't understand this case, could you please say more about when will this case happen? Why do we need msgpack here? Isn't pickle5 enough?

Isn't this a regression? Today, we can serialize `[1, 2, PyObject(obj_ref)]` fine as an x-lang object. The outer list is msgpack, the PyObject is handled by pickle5, and the ObjectRef handled by a custom serializer within pickle5. This is assuming ObjectRef is registered as a serializer plugin.

> Sure I can submit a prototype PR, I'm confident that it will pass all tests. But the code might not be simpler immediately as I mentioned above.

It is fine if the code is longer, or has more abstractions, but overall we are hoping it becomes more maintainable. For example, the pickle5 / other registration hook hacks should go away, and it should be registered in the new infrastructure.

lixin-wei commented:
> Isn't this a regression? Today, we can serialize [1, 2, PyObject(obj_ref)] fine as an x-lang object. The outer list is msgpack, the PyObject is handled by pickle5, and the ObjectRef handled by a custom serializer within pickle5. This is assuming ObjectRef is registered as a serializer plugin.

We can still do this in the new infrastructure. Since we'll offer some common container serializers.

lixin-wei commented:
Hi @ericl, prototype submitted, would you please have a look? <https://github.com/ray-project/ray/pull/25931>

There are some performance concerns to be discussed, search 'TODO'. But the overall code has passed all related tests and is ready to be reviewed.

cc @raulchen

jovany-wang commented:
> it supports buffer-based objects like numpy ndarray, pytorch tensor, pyarrow table; they can be encoded as a dict + buffers.
It sounds good to me.

> it supports GPU and other buffers
this serialization protocol allows us to transfer objects with zero-copy across languages and across frameworks (numpy -> pytorch -> jax tensor -> ...), which cannot be supported by pyarrow
For this point, I'm not sure if it's the best solution to address GPU buffer or tensors in serialization framework. What I'd like to do is that we should make sure whatever our users do, it always has a good perf.

jovany-wang commented:
Because of the lack of accessibility to [lixin-wei:serialization](https://github.com/lixin-wei/enhancements/tree/serialization) repo, I'll reopen a new PR.
